### PR TITLE
Report NFW on background thread

### LINKED ISSUE #29100: Roslyn should never report a non-fatal watson on the UI thread
_From @davkean on August 3, 2018 7:9_

I raised this issue here: https://github.com/dotnet/roslyn/issues/20101, but the reporting part got lost. I'm running into UI delays due to Roslyn reporting watsons on the UI thread. If I look around the tree, other components, such as CPS (see ProjectFaultHandlerService.HandleErrorAsync), push the work onto a background thread.

```

Frame | IncCt | IncAvg
-- | -- | --
ROOT | 1 | 49,444
ntdll.dll!_RtlUserThreadStart | 1 | 49,444
ntdll.dll!__RtlUserThreadStart | 1 | 49,444
kernel32.dll!BaseThreadInitThunk | 1 | 49,444
devenv.exe!__scrt_common_main_seh | 1 | 49,444
devenv.exe!WinMain | 1 | 49,444
devenv.exe!CDevEnvAppId::Run | 1 | 49,444
devenv.exe!util_CallVsMain | 1 | 49,444
msenv.dll!VStudioMain | 1 | 49,444
msenv.dll!VStudioMainLogged | 1 | 49,444
msenv.dll!CMsoComponent::PushMsgLoop | 1 | 49,444
msenv.dll!SCM_MsoCompMgr::FPushMessageLoop | 1 | 49,444
msenv.dll!SCM::FPushMessageLoop | 1 | 49,444
msenv.dll!CMsoCMHandler::FPushMessageLoop | 1 | 49,444
msenv.dll!CMsoCMHandler::EnvironmentMsgLoop | 1 | 49,444
msenv.dll!MainMessageLoop::ProcessMessage | 1 | 49,444
user32.dll!DispatchMessageW | 1 | 49,444
user32.dll!DispatchMessageWorker | 1 | 49,444
user32.dll!UserCallWinProcCheckWow | 1 | 49,444
user32.dll!_InternalCallWinProc | 1 | 49,444
windowsbase.dll!MS.Win32.HwndSubclass.SubclassWndProc | 1 | 49,444
windowsbase.dll!System.Windows.Threading.Dispatcher.LegacyInvokeImpl | 1 | 49,444
windowsbase.dll!System.Windows.Threading.ExceptionWrapper.TryCatchWhen | 1 | 49,444
windowsbase.dll!System.Windows.Threading.ExceptionWrapper.InternalRealCall | 1 | 49,444
windowsbase.dll!MS.Win32.HwndSubclass.DispatcherCallbackOperation | 1 | 49,444
windowsbase.dll!MS.Win32.HwndWrapper.WndProc | 1 | 49,444
windowsbase.dll!System.Windows.Threading.Dispatcher.WndProcHook | 1 | 49,444
windowsbase.dll!System.Windows.Threading.Dispatcher.ProcessQueue | 1 | 49,444
windowsbase.dll!System.Windows.Threading.DispatcherOperation.Invoke | 1 | 49,444
windowsbase.dll!MS.Internal.CulturePreservingExecutionContext.Run | 1 | 49,444
mscorlib.dll!System.Threading.ExecutionContext.Run | 1 | 49,444
mscorlib.dll!System.Threading.ExecutionContext.RunInternal | 1 | 49,444
windowsbase.dll!System.Windows.Threading.DispatcherOperation.InvokeInSecurityContext | 1 | 49,444
windowsbase.dll!System.Windows.Threading.DispatcherOperation.InvokeImpl | 1 | 49,444
windowsbase.dll!System.Windows.Threading.ExceptionWrapper.TryCatchWhen | 1 | 49,444
microsoft.codeanalysis.editorfeatures.dll!Microsoft.CodeAnalysis.Editor.Shared.Utilities.SynchronizationContextTaskScheduler.PostCallback | 1 | 49,444
mscorlib.dll!System.Threading.Tasks.TaskScheduler.TryExecuteTask | 1 | 49,444
mscorlib.dll!System.Threading.Tasks.Task.ExecuteEntry | 1 | 49,444
mscorlib.dll!System.Threading.Tasks.Task.ExecuteWithThreadLocal | 1 | 49,444
mscorlib.dll!System.Threading.ExecutionContext.Run | 1 | 49,444
mscorlib.dll!System.Threading.Tasks.Task.ExecutionContextCallback | 1 | 49,444
mscorlib.dll!System.Threading.Tasks.Task.Execute | 1 | 49,444
microsoft.visualstudio.languageservices.dll!Microsoft.VisualStudio.LanguageServices.Packaging.PackageInstallerService+<>c__DisplayClass_0.<OnWorkspaceChanged>b__ | 1 | 49,444
microsoft.visualstudio.languageservices.dll!Microsoft.VisualStudio.LanguageServices.Packaging.PackageInstallerService.ProcessBatchedChangesOnForeground | 1 | 49,444
microsoft.visualstudio.languageservices.dll!Microsoft.VisualStudio.LanguageServices.Packaging.PackageInstallerService.ProcessProjectChange | 1 | 49,444
microsoft.visualstudio.languageservices.dll!Microsoft.CodeAnalysis.ErrorReporting.WatsonReporter.Report | 1 | 49,444
microsoft.visualstudio.telemetry.dll!Microsoft.VisualStudio.Telemetry.TelemetrySession.PostEvent | 1 | 49,444
microsoft.visualstudio.telemetry.dll!Microsoft.VisualStudio.Telemetry.SessionChannel.WatsonSessionChannel.PostEvent | 1 | 49,444
microsoft.visualstudio.telemetry.dll!Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.PostWatsonReport | 1 | 49,444
mscorlib.dll!System.Threading.WaitHandle.WaitOne | 1 | 49,444
mscorlib.dll!System.Threading.WaitHandle.InternalWaitOne | 1 | 49,444
clr.dll!WaitHandleNative::CorWaitOneNative | 1 | 49,444
clr.dll!Thread::DoAppropriateWait | 1 | 49,444
clr.dll!Thread::DoAppropriateWaitWorker | 1 | 49,444
clr.dll!Thread::DoSyncContextWait | 1 | 49,444
clr.dll!MethodDescCallSite::CallTargetWorker | 1 | 49,444
clr.dll!CallDescrWorkerWithHandler | 1 | 49,444
clr.dll!CallDescrWorkerInternal | 1 | 49,444
microsoft.visualstudio.telemetry.dll!Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.NoPumpSyncContext.Wait | 1 | 49,444
kernelbase.dll!WaitForMultipleObjects | 1 | 49,444

```

_Copied from original issue: dotnet/project-system#3825_

### PR DESCRIPTION
**Pull request by genlu**
This is an attempt to fix #29100. 

@heejaechang @jasonmalinowski @davkean @AArnott 
Could you guys please take a look?

### CODE REVIEW

File: src/VisualStudio/Core/Def/Implementation/Watson/WatsonReporter.cs [hunk 1]
```diff
@@ -21,6 +24,20 @@ internal static class WatsonDisabled
...
+        /// </summary>
+        private static JoinableTaskCollection s_asyncTasks = new JoinableTaskCollection(ThreadHelper.JoinableTaskContext);
+
+        /// <summary>
+        /// Wait for started reporting tasks to finish. 
+        /// This should be called during shutdown to avoid dropping NFW.
+        /// </summary>
+        public static void WaitForPendingReporting()
+#pragma warning disable VSTHRD102 // Implement internal logic asynchronously
+            => ThreadHelper.JoinableTaskFactory.Run(async () => await s_asyncTasks.JoinTillEmptyAsync().ConfigureAwait(false));
```
Comment (AArnott) [inline src/VisualStudio/Core/Def/Implementation/Watson/WatsonReporter.cs]: tip: drop `async`, `await` and `.ConfigureAwait(false)` from this.

File: src/VisualStudio/Core/Def/Implementation/Watson/WatsonReporter.cs [hunk 2]
```diff
@@ -95,7 +112,12 @@ public static void Report(string description, Exception exception, Func<IFaultUt
             // watson and telemetry. 
             faultEvent.SetExtraParameters(exception, emptyCallstack);
 
-            TelemetryService.DefaultSession.PostEvent(faultEvent);
+            s_asyncTasks.Add(ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
```
Comment (AArnott) [inline src/VisualStudio/Core/Def/Implementation/Watson/WatsonReporter.cs]: This won't add the `JoinableTask` to the `JoinableTaskCollection` until after the first yielding await (when `RunAsync` actually returns to the caller). If you were already on a background thread when you called this, the only await in that delegate would not return and therefore you wouldn't have the protection you thought you had.
You should create a `JoinableTaskFactory s_asyncTasksFactory` field next to your `JoinableTaskCollection`, and initialize the factory based on the collection (the factory constructor takes a collection as an optional argument). Then use that factory here. That will guarantee that the JoinableTask is added to the collection even before invoking the delegate.

Be careful with static field initializers depending on one another, of course. I would be explicit about it by using a static cctor:

```cs
static WatsonReporter()
{
  s_asyncTasks = ThreadHelper.JoinableTaskContext.CreateCollection();
  s_asyncTasksFactory = ThreadHelper.JoinableTaskContext.CreateFactory(s_asyncTasks);
}
```

**Comment by jasonmalinowski** [comment]
Echoing @AArnott's concern I'm indeed not sure if this helps the root cause. Don't we also get crash dumps for the point of the NFW reporting? Moving the thread it's being invoked on means we're losing the crash context...?

**Comment by ryzngard** [comment]
Echoing my thoughts from [the issue comment](https://github.com/dotnet/roslyn/issues/29100#issuecomment-521789309) 

We will lose context when the dump is created. I think this fix will help unblock the UI thread, but can hinder our investigations. I believe there's a more reasonable approach we can take by pushing changes to the `TelemetryService.DefaultSession.PostEvent(faultEvent);` code.

**Comment by genlu** [comment]
>  Moving the thread it's being invoked on means we're losing the crash context...?

@davkean How did you guys solve this issue in CPS when switched off UI thread for reporting?

**Comment by davkean** [comment]
VS is freezing for up to 30 seconds -> 60 seconds when Roslyn takes a dump on my machine, something additional is happening other than freezing the process to take a snapshot of the process. I suspect its hitting the network somewhere . For example, there was a situation where ToString hit symbol servers, I suspect something along those lines.

**Comment by genlu** [comment]
> there was a situation where ToString hit symbol servers

I'm curious, what would trigger this?

> I suspect something along those lines.

yea, I understand your concern. I'm just asking how did you guys address the lost-context issue?

**Comment by AArnott** [comment]
In CPS (at least when I worked on it) we accepted the lost context issue as a necessary loss. We preserved the original exception and re-threw it within an exception filter such that the watson we got blamed the right callstack, but the state of the rest of the process evolved since the exception was thrown so we had to accept that.
In CPS everything is so async anyway that waiting a few more moments before taking the snapshot rarely was a problem anyway IIRC.

**Comment by tmat** [comment]
@AArnott Shouldn't `TelemetrySession.PostEvent` be asynchronous? It'd expect it to capture the process dump synchronously, to preserve the state of the stack precisely (could be an option whether precision is required or not), and then do the rest of the reporting asynchronously, including calling the `gatherEventDetails` callback. 

Apart from collecting memory dump we also want to include ServiceHub logs in the Watson report. That means enumerating files on disk. That can be done asynchronously. Unlike the stack state, the logs do not need to be an exact snapshot at the point we encounter the unexpected exception. The logs are additive so they might have more information but not less.

**Comment by tmat** [comment]
It would be even better if there was an API that captures the memory dump separately from `PostEvent`. The only thing this API would do is capturing the dump and return back some object that represents the captured data. Then we would call `PostEvent` as usual, but on background thread and `gatherEventDetails` could add the crash dump object to the report.

**Comment by AArnott** [comment]
@tmat I don't know. Ashok Kamath would probably be the best person to talk to. I don't know if he has a github account.

**Comment by sharwell** [comment]
Superseded by #44094

**Pull request closed without merging**